src/MongoDB.Driver.Core/Core/WireProtocol/Messages/Encoders/BinaryEncoders/CompressedMessageBinaryEncoder.cs

```diff
@@ -0,0 +1,257 @@
+            if (stream.Position != 0)
+            {
+                // set to start
+                stream.Position = 0;
+            }
+            var compressedMessageInfo = ReadMessage(stream);
+            var headerBytes = GetOriginalHeaderBytes(compressedMessageInfo);
+
+            var compressor = _compressorFactory.GetOrCreate(compressedMessageInfo.CompressorType);
+            var uncompressedBytes = compressor.Decompress(compressedMessageInfo.CompressedBytes);
+            using (var memoryStream = WriteUncompressedMessageIntoStream(compressedMessageInfo, uncompressedBytes))
+            {
+                var encoder = GetEncoderByOpcode(compressedMessageInfo.OriginalOpCode, memoryStream);
+                var originalMessage = encoder.ReadMessage();
+                var originalBytes = new List<byte>();
+                originalBytes.AddRange(headerBytes);
+                originalBytes.AddRange(uncompressedBytes);
+
+                return new CompressedMessage(originalMessage, originalBytes.ToArray(), compressedMessageInfo.CompressorType, _compressorFactory);
+            }
+        }
+
+        /// <summary>
+        /// Writes the message.
+        /// </summary>
+        /// <param name="message">The message.</param>
+        public void WriteMessage(CompressedMessage message)
+        {
+            Ensure.IsNotNull(message, nameof(message));
+
+            var writer = CreateBinaryWriter();
+            var stream = writer.BsonStream;
+            var messageStartPosition = stream.Position;
+
+            var originalMessageBytes = EncodeOriginalMessage(message.OriginalMessage);
+            message.SaveOriginalBytes(originalMessageBytes);
+
+            WriteHeader(stream, originalMessageBytes, message.CompressorType);
+
+            var compressor = _compressorFactory.GetOrCreate(message.CompressorType);
+            var compressedBytes = compressor.Compress(originalMessageBytes, MessageHeaderLength);
+
+            stream.Write(compressedBytes, 0, compressedBytes.Length);
+            stream.BackpatchSize(messageStartPosition);
+        }
+
+        MongoDBMessage IMessageEncoder.ReadMessage()
+        {
+            return ReadMessage();
+        }
+
+        void IMessageEncoder.WriteMessage(MongoDBMessage message)
+        {
+            WriteMessage((CompressedMessage)message);
+        }
+
+        // private methods
+        private byte[] EncodeOriginalMessage(MongoDBMessage originalMessage)
+        {
+            using (var stream = new MemoryStream())
+            {
+                var encoderFactory = new BinaryMessageEncoderFactory(stream, _encoderSettings);
+
+                var encoder = originalMessage.GetEncoder(encoderFactory);
+                encoder.WriteMessage(originalMessage);
+
+                return stream.ToArray();
+            }
+        }
+
+        private IMessageEncoder GetEncoderByOpcode(Opcode opcode, Stream stream)
+        {
```

Comment:
Also, not sure about this place. Here I want to get the encoder for an original message. Can we have here something different than `OpMsg`? If so the number of `case`s should be bigger.

Reply:
At the moment we are only expecting two types of messages back from the server:

    Opcode.Reply
    Opcode.Compressed

There is no harm in adding cases for all opcodes.

Reply:
Ok, but not sure which params should be passed into: `GetReplyMessageEncoder` and `GetInsertMessageEncoder`

Reply:
But based on the tests results, it looks like this way will work